cmake: dfu: Support extra images in DFU packages (#24637)
nrfconnect/sdk-nrf: doc/nrf/libraries/dfu/dfu_extra.rst

This flexibility enables applications to support firmware updates of additional images extending beyond the native capabilities of the |NCS|, for example, for external devices.

This CMake extension supports the following DFU delivery mechanisms:

* Multi-image binary (:file:`dfu_multi_image.bin`)

  Multi-image binary is used by applications that require downloading the DFU package containing multiple images as a single blob.
  The application is then responsible for handling the images.
  Refer to the :ref:`lib_dfu_multi_image` documentation page to see how to provide firmware implementation.

* ZIP packages (:file:`dfu_application.zip`)

  ZIP packages are used for updates through Simple Management Protocol (SMP).
  The firmware responsible for handling the images uses SMP image management, which automatically handles DFU for up to three images (MCUboot slots).
  In case you need to add another image, you must adjust the :file:`zephyr_img_mgmt.c` file.
  For more information about SMP, see the :ref:`Device Management <zephyr:device_mgmt>` documentation.

Configuration
*************

> To enable extra image support, set the :kconfig:option:`MCUBOOT_EXTRA_IMAGES` Kconfig option to a value greater than 0.

Reviewer: ```suggestion
To enable extra image support, set the :kconfig:option:`MCUBOOT_EXTRA_IMAGES` Kconfig option to a value greater than ``0``.
```

Author: Thanks, this is correct.

Usage examples
**************

Add extra binaries in your application's :file:`sysbuild.cmake`` file:

.. code-block:: cmake

   if(SB_CONFIG_MCUBOOT_EXTRA_IMAGES)
      # Include the extra DFU system
      include(${ZEPHYR_NRF_MODULE_DIR}/cmake/dfu_extra.cmake)

      # Define path to extra image
      set(ext_fw "${APP_DIR}/ext_fw.bin")

      # Create build target for the extra binary
      add_custom_target(ext_fw_target DEPENDS ${ext_fw})

      # Add extra binary to DFU packages
      dfu_extra_add_binary(
          BINARY_PATH ${ext_fw}
          NAME "ext_fw"
          DEPENDS ext_fw_target
       )
   endif()

If an additional binary needs to be handled by MCUboot as updatable slots, adjust the :kconfig:option:`SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES` Kconfig option, and make sure MCUboot partitions are defined correctly.

To automatically include the extra binary in the final :file:`merged.hex` output, define a partition with the same name as provided in the ``NAME`` parameter within your partition manager configuration.

API documentation
*****************

See a list of specific CMake functions used within the system.

dfu_extra_add_binary()

> =======================

Reviewer: ```suggestion
dfu_extra_add_binary()
======================
```

.. _lib_dfu_extra:

DFU extra image extensions
##########################

.. contents::
   :local:
   :depth: 2

The DFU extra image module provides a flexible mechanism for including extra firmware images in the DFU packages.

> This allows applications to extend the built-in DFU functionality with additional extra firmware images beyond ones supported natively in the nRF Connect SDK, for example firmware for external device.

Reviewer 2: ```suggestion
This flexibility enables applications to support firmware updates for external devices, extending beyond the native capabilities of the |NCS|.
```

This CMake extension supports the following DFU delivery mechanisms:

> * **Multi-image binary** (``dfu_multi_image.bin``)

Reviewer 2: ```suggestion
* Multi-image binary (:file:`dfu_multi_image.bin`)
```

>   Multi-image binary is used by applications that require downloading the DFU package containing multiple images as a single blob. The application is then responsible for handling the images. Refer to :ref:`lib_dfu_multi_image` to see how to provide firmware implementation.

Reviewer 2: ```suggestion
  Multi-image binary is used by applications that require downloading the DFU package containing multiple images as a single blob. 
  The application is then responsible for handling the images.
  Refer to the :ref:`lib_dfu_multi_image` documentation page to see how to provide firmware implementation.
```

> * **ZIP packages** (``dfu_application.zip``)
> 
>   ZIP packages are used for updates via SMP protocol. The firmware responsible for handling the images uses SMP image management which automatically handles DFU of up to 3 images (MCUboot slots). In case addition of yet another image is needed, user needs to adjust the ``zephyr_img_mgmt.c`` file. For more information about SMP, see :ref:`Device Management <zephyr:device_mgmt>` documentation.

Reviewer 2: ```suggestion
* ZIP packages (:file:`dfu_application.zip`)

  ZIP packages are used for updates through Simple Management Protocol (SMP). 
  The firmware responsible for handling the images uses SMP image management, which automatically handles DFU for up to three images (MCUboot slots). 
  In case you need to add another image, you must adjust the :fle:`zephyr_img_mgmt.c` file. 
  For more information about SMP, see the :ref:`Device Management <zephyr:device_mgmt>` documentation.
```

Configuration
*************

To enable extra image support, set the :kconfig:option:`SB_CONFIG_DFU_EXTRA_BINARIES` Kconfig option.

> Usage Examples

Reviewer 2: ```suggestion
Usage examples
```

*************

> To enable extra image support, set the :kconfig:option:`MCUBOOT_EXTRA_IMAGES` Kconfig option to value greater than 0.

Reviewer 2: ```suggestion
To enable extra image support, set the :kconfig:option:`MCUBOOT_EXTRA_IMAGES` Kconfig option to a value greater than 0.
```

.. code-block:: cmake

   if(SB_CONFIG_DFU_EXTRA_BINARIES)
      # Include the extra DFU system
      include(${ZEPHYR_NRF_MODULE_DIR}/cmake/dfu_extra.cmake)

      # Define path to external firmware
      set(ext_fw "${APP_DIR}/ext_fw.signed.bin")

      # Create build target for the external binary
      add_custom_target(ext_fw_target
         DEPENDS ${ext_fw}
         COMMENT "External firmware dependency"
      )

      # Add extra binary to DFU packages
      dfu_extra_add_binary(
          IMAGE_ID 1
          BINARY_PATH ${ext_fw}
          DEPENDS ext_fw_target
       )
   endif()

> In case additional binary needs to be handled by MCUboot as updatable slots, adjust the :kconfig:option:`SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES` and make sure MCUboot partitions are defined correctly.

Reviewer 2: ```suggestion
In case additional binary needs to be handled by MCUboot as updatable slots, adjust the :kconfig:option:`SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES` Kconfig option, and make sure MCUboot partitions are defined correctly.
```

If an additional binary needs to be handled by MCUboot as updatable slots, adjust the :kconfig:option:`SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES` Kconfig option, and make sure MCUboot partitions are defined correctly.

> To automatically include the extra binary in the final :file:`merged.hex` output, define a partition with the same name as provided in the ``NAME`` parameter within your partition manager configuration

Reviewer 2: ```suggestion
To automatically include the extra binary in the final :file:`merged.hex` output, define a partition with the same name as provided in the ``NAME`` parameter within your partition manager configuration.
```

API documentation
*****************

See a list of specific CMake functions used within the system.

dfu_extra_add_binary()
=======================

Adds an extra binary to DFU multi-image binary and ZIP packages.

Refer to the syntax of the binary:

.. code-block:: cmake

   dfu_extra_add_binary(
     BINARY_PATH <path>
     [NAME <name>]
     [VERSION <version>]
     [PACKAGE_TYPE <type>]
     [DEPENDS <target1> [<target2> ...]]
   )

You can adjust the following parameters:

* ``BINARY_PATH`` - Path to the binary file that will be included in the package.
  The path can be absolute or relative to the build directory.

* ``NAME`` - Optional name for the binary file in packages.
  If not provided, it defaults to the basename of ``BINARY_PATH``.

* ``VERSION`` - Optional image version string (for signing).

* ``PACKAGE_TYPE`` - Optional package type selection: ``"zip"``, ``"multi"``, or ``"all"`` (default).
  Controls which package types include this binary:

  * ``"zip"`` - Include only in ZIP packages
  * ``"multi"`` - Include only in multi-image binary

>   * ``"all"`` - Include in both package types

Reviewer 2: ```suggestion
  * ``"zip"`` - Included only in ZIP packages
  * ``"multi"`` - Included only in multi-image binary
  * ``"all"`` - Included in both package types
```

* ``DEPENDS`` - Optional list of CMake targets that must be built before this extra binary is available.
  This ensures the correct build sequence.

Notes:

* Image IDs are automatically assigned based on the order binaries are added.

>   For more information about image ID assignment, see :ref:`sysbuild_assigned_images_ids`.

Reviewer 2: ```suggestion
Image IDs are automatically assigned based on the order binaries are added.
For more information about image ID assignment, see :ref:`sysbuild_assigned_images_ids`.
```

> * :ref:`sysbuild_assigned_images_ids`

Author: I have removed this section. This library is pure CMake library, and don't have typical C API.
